### PR TITLE
First Version of the MKR library for Ubidots

### PR DESCRIPTION
**Pull request by crisap94**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,143 @@
...
+4. Select the MKR1010 board from **Tools > Board > Arduino MKR WiFI 1010**.
+5. Download the **Ubidots library** [here](https://github.com/ubidots/ubidots-ArduinoMKR/archive/master.zip).
+6. Now, click on **Sketch -> Include Library -> Add .ZIP Library**.
+7. Select the .ZIP file of Ubidots and then "**Accept**" or "**Choose**".
+8. Close the Arduino IDE and open it again.
+
+**NOTE:** To be able to use this device with TCP over TLS1.2 you must add manually the certificate from Ubidots to the module following the section **Certificate uploading** from this [guide](https://www.arduino.cc/en/Tutorial/FirmwareUpdater), and add the next line required for the process in the *Add Domain* button.
+
+`industrial.api.ubidots.com:9812`
+
```
Comment (jotathebest) [inline README.md:31]: Add instructions to upload the TLS certificates needed to connect with Ubidots

File: src/UbiBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
+
+/*
+Copyright (c) 2013-2019 Ubidots.
```
Comment (jotathebest) [inline src/UbiBuilder.cpp]: 2020

File: src/UbiBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+UbiProtocol *builderTcp() {
+  UbiProtocol *tcpInstance =
+      new UbiTCP(_host, UBIDOTS_TCPS_PORT, USER_AGENT, _token);
+  return tcpInstance;
+}
+
+UbiProtocol *builderHttp() {
+  UbiProtocol *httpInstance =
+      new UbiHTTP(_host, UBIDOTS_HTTPS_PORT, USER_AGENT, _token);
```
Comment (jotathebest) [inline src/UbiBuilder.cpp]: set your linter to 120 lines

File: src/UbiBuilder.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+typedef UbiProtocol *(*builderProtocol)(void);
+
+namespace {
+const char *_host;
+const char *_token;
+} // namespace
+
+class UbiBuilder {
+public:
+  explicit UbiBuilder(const char *host, const char *token,
```
Comment (jotathebest) [inline src/UbiBuilder.h]: 120 lines

File: src/UbiConstants.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+const char UBIDOTS_INDUSTRIAL_IP[] = "169.55.61.243";
+const char *const USER_AGENT = "UbidotsArduinoMKR1000/1.0.0";
+const int UBIDOTS_HTTPS_PORT = 443;
+const int UBIDOTS_TCP_PORT = 9012;
+const int UBIDOTS_TCPS_PORT = 9812;
+const uint8_t MAX_VALUES = 10;
+const float ERROR_VALUE = -3.4028235E+8;
+const int MAX_BUFFER_SIZE = 700;
+static UbiServer UBI_INDUSTRIAL = "industrial.api.ubidots.com";
+const int SUPPORTED_PROTOCOLS = 3;
```
Comment (jotathebest) [inline src/UbiConstants.h]: change this to NUMBER_OF_SUPPORTED_PROTOCOLS

File: src/UbiConstants.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
+
+/*
+Copyright (c) 2013-2018 Ubidots.
```
Comment (jotathebest) [inline src/UbiConstants.h:3]: 2020

File: src/UbiHttp.cpp [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+#include "UbiHttp.h"
+
+#include "UbiUtils.h"
+
+/**************************************************************************
+ * Overloaded constructors
+ ***************************************************************************/
+
+UbiHTTP::UbiHTTP(const char *host, const int port, const char *user_agent,
```
Comment (jotathebest) [inline src/UbiHttp.cpp]: set the user_agent as a default parameter as it is not mandatory for sending data, const char *user_agent=USER_AGENT. I would say, that it should not be even a parameter for this object

File: src/UbiHttp.cpp [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+    }
+  }
+
+  free(message);
+  free(path);
+
+  double value = _parseServerAnswer();
+  _client_https_ubi.flush();
+  _client_https_ubi.stop();
+  // return 0;
```
Comment (jotathebest) [inline src/UbiHttp.cpp]: erase this

File: src/UbiHttp.cpp [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+  uint8_t length = UbiUtils::hexadecimalToDecimal(_charLength);
+
+  if (_debug) {
+    Serial.print(F("Length: "));
+    Serial.println(length);
+  }
+
+  char *_charValue = (char *)malloc(sizeof(char) * length + 1);
+
+  readServerAnswer(_charValue);
```
Comment (jotathebest) [inline src/UbiHttp.cpp]: do not read twice the server's answer

File: src/UbiHttp.cpp [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+
+double UbiHTTP::_parseServerAnswer() {
+
+  /**
+   * @param _charResponseLength 3 is the maximun amount of digits for the
+   * length to have
+   */
+  char *_charLength = (char *)malloc(sizeof(char) * 3);
+
+  readServerAnswer(_charLength);
```
Comment (jotathebest) [inline src/UbiHttp.cpp]: what happens here if the variable or the device does not existe in the account?

File: src/UbiHttp.cpp [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+                            strlen(device_label) + strlen(variable_label);
+  return endpointLength;
+}
+
+/**
+ * @arg response [Mandatory] Pointer to store the server's answer
+ */
+
+void UbiHTTP::readServerAnswer(char *_serverResponse) {
+  bool _firstChar = true;
```
Comment (jotathebest) [inline src/UbiHttp.cpp:302]: refact this method, there are too many ifs that makes hard-reading the logic of the method

File: src/UbiHttp.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright (c) 2013-2019 Ubidots.
```
Comment (jotathebest) [inline src/UbiHttp.h]: 2020

File: src/UbiHttp.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+*/
+
+#ifndef _UbiHttp_H_
+#define _UbiHttp_H_
+
+#include "UbiProtocol.h"
+
+class UbiHTTP : public UbiProtocol {
+public:
+  UbiHTTP(const char *host, const int port, const char *user_agent,
```
Comment (jotathebest) [inline src/UbiHttp.h]: 120 lines

File: src/UbiProtocol.h [hunk 14]
```diff
@@ -0,0 +1,102 @@
+/*
+Copyright (c) 2013-2018 Ubidots.
```
Comment (jotathebest) [inline src/UbiProtocol.h:2]: 2020

File: src/UbiProtocol.h [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+Inc
+@jotathebest at github: https://github.com/jotathebest
+@crisap94 at github: https://github.com/crisap94
+*/
+
+#ifndef _UbiProtocol_H_
+#define _UbiProtocol_H_
+
+#include <WiFiNINA.h>
+#include <WiFiUdp.h>
```
Comment (jotathebest) [inline src/UbiProtocol.h:29]: why do you need udp here?

File: src/UbiProtocol.h [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  /**
+   * Makes available debug traces
+   */
+
+  inline void setDebug(bool debug) { _debug = debug; }
+
+private:
+  bool reconnectionRoutine() {}
```
Comment (jotathebest) [inline src/UbiProtocol.h]: why two reconnect routines? one public and the other one private?

File: src/UbiProtocolHandler.cpp [hunk 17]
```diff
@@ -0,0 +1,278 @@
+/*
+Copyright (c) 2013-2018 Ubidots.
```
Comment (jotathebest) [inline src/UbiProtocolHandler.cpp:2]: 2020

File: src/UbiProtocolHandler.cpp [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+ * @arg dot_timestamp_seconds [optional] Dot timestamp in seconds, usefull for
+ * datalogger. Default NULL
+ * @arg dot_timestamp_millis [optional] Dot timestamp in millis to add to
+ * dot_timestamp_seconds, usefull for datalogger.
+ */
+
+void UbiProtocolHandler::add(const char *variable_label, float value,
+                             char *context, unsigned long dot_timestamp_seconds,
+                             unsigned int dot_timestamp_millis) {
+  _dirty = true;
```
Comment (jotathebest) [inline src/UbiProtocolHandler.cpp]: delete this

File: src/UbiProtocolHandler.h [hunk 19]
```diff
@@ -0,0 +1,63 @@
+/*
+Copyright (c) 2013-2019 Ubidots.
```
Comment (jotathebest) [inline src/UbiProtocolHandler.h]: 2020

**Review by jotathebest — CHANGES_REQUESTED** [review]

File: src/UbiProtocol.h (hunk 15 above)
Comment (crisap94) [reply src/UbiProtocol.h:29]: There is no UDP client inside the WiFiNINA library, WifiNINA package comes with the two libraries in separated files

File: src/UbiProtocol.h (hunk 16 above)
Comment (crisap94) [reply src/UbiProtocol.h]: This is an outdated version of the file, check again the code 'cause i don't have that method not even defined.

File: src/UbiHttp.cpp (hunk 10 above)
Comment (crisap94) [reply src/UbiHttp.cpp]: I did not think about it, although it will respond to the user with Internal Server Error message line 328.

**Review by jotathebest — APPROVED** [review]

**Pull request merged**